### PR TITLE
♻️ Enable passing type-checking on src/context

### PR DESCRIPTION
**Pull request by rcebulko**
Part of #34096

### CODE REVIEW

**Review by danielrozenberg — APPROVED** [review]
Approval for build-system/*

File: src/context/values.js [hunk 1]
```diff
@@ -37,35 +38,48 @@ const Pending = {
...
- *   ping: function(boolean),
- *   pingDep: !Array<function(*)>,
- *   pingParent: ?function(*),
- * }}
+ * @interface
+ * @template T
+ * @template DEP
  */
-let UsedDef;
+function UsedDef() {}
```
Comment (samouri) [inline src/context/values.js:56]: FMI: did you have to make this format change away from typedef due to incompatibility with templates?

File: src/context/values.js [hunk 2]
```diff
@@ -174,12 +195,12 @@ export class Values {
...
    *
-   * @param {!ContextProp} prop
+   * @param {!ContextPropDef} prop
    * @return {boolean}
    */
   has(prop) {
-    const inputsByKey = this.inputsByKey_;
-    return !!inputsByKey && inputsByKey.has(prop.key);
+    const {key} = prop;
+    return !!this.inputsByKey_?.has(key);
```
Comment (samouri) [inline src/context/values.js:203]: ignorable super nit: very slightly nicer as `prop.key`

**Review by samouri — APPROVED** [review]
Really impressive. This is not easy code to work with

File: src/context/values.js (hunk 1 above)
Comment (rcebulko) [reply src/context/values.js:56]: Yes

File: src/context/values.js (hunk 2 above)
Comment (rcebulko) [reply src/context/values.js:203]: I agree! I think I assumed it was destructured in a 2nd or 3rd pass. One-line it shall be!

**Comment by rcebulko** [comment]
Merging since @caroqliu  is OOO all week, but pinging for retroactive visibility

**Pull request merged**